rfcs/0000-support-injected-sidecars.md

```diff
@@ -0,0 +1,124 @@
+       leading the reconciler to believe the build has failed. To circumvent this we will need allow pods to
+       fail with the reason `DeadlineExceeded` as long the completion container has successfully completed.
+5. We will also need a way to delay the start of a build until the sidecars are ready in the event that they are
+   required for network communication  (i.e istio)
+    1. To do this we will have the prepare step wait on a file passed in through
+       the [kubernetes downward api][downward-api]. This file can be provided from a `build ready` annotation that is
+       added to the pod when by the build reconciler when all the containers in the pod are in a ready state.
+
+[kpack-spike-build-waiter]: https://github.com/pivotal/kpack/blob/execute-builds-in-regular-containers/cmd/build-waiter/main.go
+
+[step-modifier]: https://github.com/pivotal/kpack/blob/44257a70c2c9e9ac5703eadc871c7c9be22cfadc/pkg/apis/build/v1alpha2/build_pod.go#L181
+
+[downward-api]: https://kubernetes.io/docs/concepts/workloads/pods/downward-api/
+
+## Complexity
+
+* Our current build logging relies heavily on the fact that our builds exist in init containers so that logic will
+  need to be reworked to support this change.
+* We would need to create an implementation specific to windows builds as well if windows support is needed
+
+## Prior Art
+
+### 1. tekton
+
+https://github.com/tektoncd/pipeline/tree/main/cmd/entrypoint#waiting-for-sidecars
+
+### 2. kpack spike
+
+https://github.com/pivotal/kpack/pull/1019
+
+## Alternatives
+
+* We can create tekton [`TaskRuns`][tekton-task-run] to execute each step of the build since tekton has come up with
+  their own way to solve this problem which
+  involves [replacing the sidecar container images with a "noop" image][tekton-stop-sidecar].
+    * Note: Tekton currently will only stop sidecars (including the istio proxy) if at least one sidecar is defined in
+      the `TaskSpec`.
+* We can build the Tekton entrypoint code as an image to use in kpack.
+* We could execute all build steps in one standard container, similar to [Pack's][pack-build-docs] `pack build` command
+  with the `--trust-builder` flag.
+    * This would involve us likely extending the `build-init` binary functionality to execute the lifecycle steps after
+      downloading the source.
+    * We would also have to modify how we build builder images to include this binary as well as the lifecycle
+
+[pack-build-docs]: https://buildpacks.io/docs/tools/pack/cli/pack_build/
+
+[tekton-task-run]: https://tekton.dev/docs/pipelines/taskruns/#overview
+
+[tekton-stop-sidecar]: https://github.com/tektoncd/pipeline/blob/210dbe8965ab5fa9c0ca53e164b86d441899b763/pkg/reconciler/taskrun/taskrun.go#L240-L283
+
+## Risks
+
+* Because all the containers will be running at once, the pod will request more resources than it does today as due to
```

Comment:
Can you please explain what do you mean by "all the containers will be running at once"?
Do you mean the containers that run the lifecycle stages? Because I'm not sure it's possible.

Reply:
The containers will all be started (as opposed to init containers which don't start until it is their turn), but the build waiter logic will prevent the actual lifecycle command from executing until the previous step completes. The container will just sit and wait.

Reply:
ohh, and this should solve the network issue that exists with the init containers, right?